### PR TITLE
Improve promString performance

### PR DESCRIPTION
**Pull request by kgeckhart**
**Background**
I recently took a heap dump + a CPU profile to see where we were spending a lot of time/resources. Here is a screenshot of the flame graph for the `UpdateMetrics` call which is the main entrypoint.
image

The big thing I noticed is how much the calls to `promString` standout. They account for 3.13s of the total 3.44s runtime of the `UpdateMetrics`function according to top
image

`promString` calls two functions, first `splitString` and then `sanitize`. `splitString` **always** complies a regex `([a-z0-9])([A-Z])` to insert `.` in between the capture groups. `sanitize` **always** creates a new `strings.NewReplacer` to replace various strings with _ to be compatible with prom. If we look at the top 10 flat users of memory on an instance which has been running for ~5 day at least 5 out of 10 are related to creating the new instances (more on the top user later)
image

**Actions Taken**

* `sanitize`
  * now uses a global`strings.NewReplacer`because it's safe for concurrent use
  * Initializing the replacer when it's first used, however many times, accounted for 0.78s of the 3.44s request time
* `splitString`
  * The safest thing to do is make the compiled regex global which will cut some CPU + memory
  * But the top flat allocator of memory comes from running the regex itself and the regex running counts for 1.6s of the total 3.44s request time captured
  * Since the regex isn't super complicated I chose to replace the regex with looping through the string instead
  * Benchmark results (code in prometheus_test.go) show really good results vs original and global
image
  * There's also a minor change in loop version to insert `_` instead of `.`. Regex replacement treats `_` as part of a capture group so `.` was used instead. The `.` is then immediately replaced with `_` when `sanitize` is called which is just extra work when not using the regex.

If this all seems reasonable I'll add some more tests + remove the benchmark code so it doesn't go stale.

### CODE REVIEW

File: pkg/prometheus.go [hunk 1]
```diff
@@ -41,6 +41,23 @@ var (
 		Name: "yace_cloudwatch_ec2api_requests_total",
 		Help: "Help is not implemented yet.",
 	})
+
+	replacer = strings.NewReplacer(
```
Comment (matthewnolf) [inline pkg/prometheus.go]: Nit - use var blocks to group related variables. Replacer isn't related to the prometheus metrics, so put into it's own var block (or inline) outside of this one

**Review by matthewnolf — COMMENTED** [review]
Thanks for the investigation @kgeckhart - this is really nice to see and gives some good insight!

+1 for moving the replacer to package level - makes sense to me.

For the change to `splitString`, could we benchmark the difference once we move the `MustCompile` call out to package level?

Reason I mention it, is if possible I'd like to use stdlibrary functions instead of our own implementations - and so if the difference equates to <=1ms, we may be optimising something unnecessarily.

**Comment by cristiangreco** [comment]
Great finding! 🚀 

I'm wondering whether it makes sense to split this PR in two parts? The change to `sanitize` seems self-contained and well understood, but `splitString` might involve a bit more discussion apparently?

**Comment by kgeckhart** [comment]
I have no problem splitting these up as I think the two global changes alone should have a solid impact 🤞 

> Reason I mention it, is if possible I'd like to use stdlibrary functions instead of our own implementations - and so if the difference equates to <=1ms, we may be optimising something unnecessarily.

I did find it odd that the benchmark via go test wasn't a lot higher given that in the CPU profile showed regexp.ReplaceAllString was 1.6s. Either way, if ReplaceAllString is still heavy the CPU profile will show it a lot more after the other changes.

**Review by matthewnolf — APPROVED** [review]
LGTM 👍

**Review by cristiangreco — APPROVED** [review]

**Pull request merged**